codec/aminojson/json_marshal.go

```diff
@@ -0,0 +1,247 @@
+package aminojson
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	cosmos_proto "github.com/cosmos/cosmos-proto"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"cosmossdk.io/api/amino"
+)
+
+type JSONMarshaller interface {
+	MarshalAmino(proto.Message) ([]byte, error)
+}
+
+type AminoJSON struct {
+	// maps cosmos_proto.scalar -> zero value factory
+	zeroValues      map[string]func() protoreflect.Value
+	messageEncoders map[string]func(message protoreflect.Message) protoreflect.Value
+}
+
+func NewAminoJSON() AminoJSON {
+	aj := AminoJSON{
+		zeroValues: map[string]func() protoreflect.Value{
+			"cosmos.Dec": func() protoreflect.Value {
+				return protoreflect.ValueOfString("0")
+			},
+		},
+		messageEncoders: map[string]func(message protoreflect.Message) protoreflect.Value{
```

Comment:
Nice. We need a way for app developers to insert custom message encoders too.

Something like
```go
func (aj AminoJSON) DefineMessageEncoding(string, func...) {}
```

CosmWasm has 1 or 2 custom `amino.message_encoding` tags.